classes/metric/concurrent_users_metrics.php

```diff
@@ -0,0 +1,134 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_cloudmetrics\metric;
+
+use cltr_database\collector;
+
+/**
+ * Metric class for new users.
+ *
+ * @package    metric_foobar
+ * @author     Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @copyright  2022, Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class concurrent_users_metrics extends builtin {
+    /**
+     * The metric's name.
+     *
+     * @return string
+     */
+    public function get_name(): string {
+        return 'concurrentusers';
+    }
+
+    /**
+     * The metric type.
+     *
+     * @return int
+     */
+    public function get_type(): int {
+        return manager::TYPE_GAUGE;
+    }
+
+    /**
+     * Retrieves the metric.
+     *
+     * @param  $backfilledrecord If data is backfilled, record is passed here.
+     * @return metric_item
+     */
+    public function get_metric_item($record = null): metric_item {
+        return new metric_item($this->get_name(), $record->time, $record->value, $this);
```

Comment:
How is this supposed to work during normal metric gathering?
What happens when $record is null?

Reply:
If possible I would like to discuss this in a bbb maybe your Monday or Tuesday morning (Sunday or Monday evening for me) ?

Reply:
Tues morning (Aussie time). If you can specify a specific time, I can make sure I am here for it. Any time is fine except for the 10:00 meeting.

Reply:
That is, I have meetings between 10:00 and 10:30 (Sydney time). Any other time is fine.

Reply:
hi @marcghaly 

I have a few concerns around the public interface, and there is a lot of overlap with issue #73 which @jaypha was working on here:

https://github.com/catalyst/moodle-tool_cloudmetrics/pull/83

That pr isn't ready but what I've proposed just now with Jason is that the public interface for normal metrics is changed to something 

get_metric_item($starttime, $finishtime, $periodlength)

This is because in reality all metrics are retrospective, they just might be a few seconds later due to the latency of running each cron process and we need to know the exact timeperiod that the data should be have been for (even if the metric itself ignores it and just gets 'now')

Once that is in place the retrospective metrics should end up very similar something like:

get_metric_items($starttime, $finishtime, $periodlength)

The only difference is that the start time and finish time will be some distant past timestamps and we are expecting an array of them. We also need to consider the problem when the number of metric items returned is vast, and we might need to limit it and return what the time period was that we are returning data for which will be a subset of what was requested. Then it can be iteratively called to progressively backfill data in smaller chunks over time and the calling code would need to manage what has been processed or not already.

So I think we should land the first PR first so you don't have a moving target. Jason will get onto that in the next day or say.

I would also still like to meet, ideally all 3 of us if we can make it work, half as a meet and greet and also to iron out any last bits on this. I'll add a couple other comments in the pr here but many of them will be redundant when this gets rewritten on top of the other pr